README.md
# Colored JSON

Convert JSON string into _customizable_ widget.

## Getting Started

`ColoredJson` is a stateless widget that produces a structured view of JSON string provided. It can be used like any other flutter widget.
It's fully customizable where you can set different colors to different datatypes including colon, comma and brackets. Set indent length. Use text style of your choice. You can also change the background color.
|Defaut View|Customised View 1|Customised View 2|
|--|--|--|
| <img src="https://i.ibb.co/5hPyLt4/default.jpg" alt="Default View"/> | <img src="https://i.ibb.co/3vnm3xr/customized.jpg" alt="Customized View 1" >|<img src="https://i.ibb.co/rk10303/flexible.jpg" alt="Customized View 2" />|

## Usage

- Pass a JSON string to the `data` parameter of `ColoredJson` Widget

### Simple Usage

```dart
const String responseBody = '''
{
  'id' : 1,
  'name' : 'Colored Json',
  'isFlutterAwesome' : true,
}
''';

ColoredJson(
 data: responseBody,
)

```

### Full Usage

```dart
final dartMap = {
  "id": 24,
  "name": "Manthan Khandale",
  "score": 7.6,
  "socials": null,
  "hobbies": [
    "Music",
    "Filmmaking",
  ],
  "isFlutterCool": true,
};

ColoredJson(
 data: jsonEncode(dartMap),
 indentLength: 5,
 keyColor: Colors.green,
 backgroundColor: Colors.black,
 boolColor: Colors.white,
 nullColor: Colors.redAccent,
 stringColor: Colors.cyan,
 curlyBracketColor: Colors.yellow,
 doubleColor: Colors.deepOrange,
 squareBracketColor: Colors.amber,
 commaColor: Colors.yellow,
 colonColor: Colors.purple,
 intColor: Colors.lime,
 textStyle: const TextStyle(
  fontSize: 20,
  letterSpacing: 2,
  ),
 )
```

| Parameter           | Info                                 |
| ------------------- | ------------------------------------ |
| **data** (required) | A **JSON** string to be formatted    |
| keyColor            | Color of Json **Key**                |
| intColor            | Color of **integer** type value      |
| doubleColor         | Color of **decimal** type value      |
| stringColor         | Color of **string** type value       |
| boolColor           | Color of **boolean** type value      |
| nullColor           | Color of **null** type value         |
| curlyBracketColor   | Color of **curly** brackets {}       |
| squareBracketColor  | Color of **square** brackets []      |
| commaColor          | Color of **comma** ( , )             |
| colonColor          | Color of **colon** ( : )             |
| backgroundColor     | **Background** color of widget       |
| textStyle           | **Textstyle** of texts in the widget |
| indentLength        | **Indent** length                    |

## Upcoming Tasks

- Code shrink feature
- Add Inner Spacing parameter
- Add tests


CHANGELOG.md
# 1.1.2

- Minor typo fixes

# 1.1.1

- Screenshots added

# 1.1.0

- Indent issue fix on new line
- Typo fixes

# 1.0.0

- Initial Release


test/preety_json_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:preety_json/preety_json.dart';

void main() {
  //todo add tests
}


test/pretty_json_test.dart
void main() {
  //todo add tests
}


lib/preety_json.dart
library preety_json;

import 'package:flutter/material.dart';
import 'dart:convert';

late int _globalTabLength;

late Color _globalCommaColor;

late Color _globalKeyColor;

late Color _globalColonColor;

class PreetyJson extends StatelessWidget {
  PreetyJson({
    Key? key,
    required this.data,
    this.backgroundColor = Colors.transparent,
    this.stringColor = Colors.green,
    this.nullColor = Colors.red,
    this.intColor = Colors.purple,
    this.boolColor = Colors.pink,
    this.doubleColor = Colors.deepPurple,
    this.curlyBracketColor = Colors.black,
    this.squareBracketColor = Colors.black,
    this.textStyle = const TextStyle(),
    this.talL = 4,
    this.keyColor = Colors.blue,
    this.commaColor = Colors.black,
    this.colonColor = Colors.deepPurple,
  }) : super(key: key) {
    _globalTabLength = talL;
    _globalCommaColor = commaColor;
    _globalColonColor = colonColor;
    _globalKeyColor = keyColor;
  }

  /// Json String
  final dynamic data;

  /// Background color of the Widget
  final Color backgroundColor;

  /// Color of value of type [String]
  final Color stringColor;

  /// Color of value of type null
  final Color nullColor;

  /// Color of value of type [int]
  final Color intColor;

  /// Color of value of type [bool]
  final Color boolColor;

  /// Color of value of type [double]
  final Color doubleColor;

  /// Color of Curly brackets
  final Color curlyBracketColor;

  /// Color of Curly brackets
  final Color squareBracketColor;

  /// Textstyle of parsed json data
  final TextStyle textStyle;

  /// Tab length in spaces, default to 4
  final int talL;

  /// Color of Json Color Key
  final Color keyColor;

  /// Color of comma
  final Color commaColor;

  /// Color of color
  final Color colonColor;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: backgroundColor,
      textStyle: textStyle,
      child: _preetyView(
        jsonString: data,
        indent: 0,
      ),
    );
  }

  Widget _preetyView({
    required dynamic jsonString,
    required int indent,
  }) {
    final decodedData = jsonDecode(jsonString);
    if (decodedData is Map) {
      return _parseMap(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else if (decodedData is List) {
      return _parseList(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else {
      throw 'Not Correct type only List or Map allowed';
    }
  }

  Widget _parseMap({
    required Map data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.entries.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{}',
          valueColor: curlyBracketColor,
          showComma: showComma,
        );
      } else {
        return Text(
          '{}',
          style: TextStyle(color: curlyBracketColor),
        );
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}{',
            style: TextStyle(color: curlyBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{',
          valueColor: curlyBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int i = 0; i < data.entries.length; i++) {
        final jKey = data.entries.toList()[i].key;
        final value = data.entries.toList()[i].value;

        final bool showComma = i != data.entries.length - 1;

        if (value is String) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: '"$value"',
                valueColor: stringColor,
                showComma: showComma),
          );
        } else if (value is int) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: value.toString(),
                valueColor: intColor,
                showComma: showComma),
          );
        } else if (value is double) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: doubleColor,
              showComma: showComma,
            ),
          );
        } else if (value is bool) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: boolColor,
              showComma: showComma,
            ),
          );
        } else if (value is List) {
          wigs.add(_parseList(
            data: value,
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value is Map) {
          wigs.add(_parseMap(
            data: (value),
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value == null) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: 'null',
              valueColor: nullColor,
              showComma: showComma,
            ),
          );
        }
      }
      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: '}', style: TextStyle(color: curlyBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }

  Widget _parseList({
    required List<dynamic> data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[]',
          valueColor: squareBracketColor,
          showComma: showComma,
        );
      } else {
        return Text('[]', style: TextStyle(color: squareBracketColor));
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}[',
            style: TextStyle(color: squareBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[',
          valueColor: squareBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int j = 0; j < data.length; j++) {
        final i = data[j];
        final showComma = j != data.length - 1;

        if (i is String) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: '"$i"',
            valueColor: stringColor,
            showComma: showComma,
          ));
        } else if (i is int) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: intColor,
            showComma: showComma,
          ));
        } else if (i is double) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: doubleColor,
            showComma: showComma,
          ));
        } else if (i is bool) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: boolColor,
            showComma: showComma,
          ));
        } else if (i is List) {
          wigs.add(_parseList(
            data: i,
            indent: newIndent,
            showComma: showComma,
          ));
        } else if (i is Map) {
          wigs.add(_parseMap(
              data: i as Map<String, dynamic>,
              indent: newIndent,
              showComma: showComma));
        } else if (i == null) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: 'null',
            valueColor: nullColor,
            showComma: showComma,
          ));
        }
      }

      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: ']', style: TextStyle(color: squareBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }
}

class _KeyValueWidget extends StatelessWidget {
  _KeyValueWidget({
    Key? key,
    required this.indent,
    this.jsonKey,
    required this.jsonValue,
    required this.valueColor,
    required this.showComma,
  }) : super(key: key);

  final int indent;
  final String? jsonKey;
  final String jsonValue;
  final Color valueColor;

  final Color keyColor = _globalKeyColor;

  final bool showComma;

  @override
  Widget build(BuildContext context) {
    return Text.rich(TextSpan(
      children: [
        TextSpan(text: _spaces(indent)),
        if (jsonKey != null)
          TextSpan(text: '"$jsonKey"', style: TextStyle(color: keyColor)),
        if (jsonKey != null)
          TextSpan(text: ' : ', style: TextStyle(color: _globalColonColor)),
        TextSpan(text: jsonValue, style: TextStyle(color: valueColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor)),
      ],
    ));
  }
}

String _spaces(int multiple) {
  return ' ' * _globalTabLength * multiple;
}


lib/pretty_json.dart
library pretty_json;

import 'package:flutter/material.dart';
import 'dart:convert';

late int _globalTabLength;

late Color _globalCommaColor;

late Color _globalKeyColor;

late Color _globalColonColor;

class PrettyJson extends StatelessWidget {
  PrettyJson({
    Key? key,
    required this.data,
    this.backgroundColor = Colors.transparent,
    this.stringColor = Colors.green,
    this.nullColor = Colors.red,
    this.intColor = Colors.purple,
    this.boolColor = Colors.pink,
    this.doubleColor = Colors.deepPurple,
    this.curlyBracketColor = Colors.black,
    this.squareBracketColor = Colors.black,
    this.textStyle = const TextStyle(),
    this.indentLength = 4,
    this.keyColor = Colors.blue,
    this.commaColor = Colors.black,
    this.colonColor = Colors.deepPurple,
  }) : super(key: key) {
    _globalTabLength = indentLength;
    _globalCommaColor = commaColor;
    _globalColonColor = colonColor;
    _globalKeyColor = keyColor;
  }

  /// Json String
  final String data;

  /// Background color of the Widget
  final Color backgroundColor;

  /// Color of value of type [String]
  final Color stringColor;

  /// Color of value of type null
  final Color nullColor;

  /// Color of value of type [int]
  final Color intColor;

  /// Color of value of type [bool]
  final Color boolColor;

  /// Color of value of type [double]
  final Color doubleColor;

  /// Color of Curly brackets
  final Color curlyBracketColor;

  /// Color of Curly brackets
  final Color squareBracketColor;

  /// Textstyle of parsed json data
  final TextStyle textStyle;

  /// Indent length in spaces, default to 4
  final int indentLength;

  /// Color of Json Color Key
  final Color keyColor;

  /// Color of comma
  final Color commaColor;

  /// Color of color
  final Color colonColor;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: backgroundColor,
      textStyle: textStyle,
      child: _prettyView(
        jsonString: data,
        indent: 0,
      ),
    );
  }

  /// Takes JSON string and initial indent
  Widget _prettyView({
    required String jsonString,
    required int indent,
  }) {
    final decodedData = jsonDecode(jsonString);
    if (decodedData is Map) {
      return _parseMap(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else if (decodedData is List) {
      return _parseList(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else {
      throw 'Not Correct type only List or Map allowed';
    }
  }

  /// Creates a Map View
  Widget _parseMap({
    required Map data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.entries.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{}',
          valueColor: curlyBracketColor,
          showComma: showComma,
        );
      } else {
        return Text(
          '{}',
          style: TextStyle(color: curlyBracketColor),
        );
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}{',
            style: TextStyle(color: curlyBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{',
          valueColor: curlyBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int i = 0; i < data.entries.length; i++) {
        final jKey = data.entries.toList()[i].key;
        final value = data.entries.toList()[i].value;

        final bool showComma = i != data.entries.length - 1;

        if (value is String) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: '"$value"',
                valueColor: stringColor,
                showComma: showComma),
          );
        } else if (value is int) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: value.toString(),
                valueColor: intColor,
                showComma: showComma),
          );
        } else if (value is double) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: doubleColor,
              showComma: showComma,
            ),
          );
        } else if (value is bool) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: boolColor,
              showComma: showComma,
            ),
          );
        } else if (value is List) {
          wigs.add(_parseList(
            data: value,
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value is Map) {
          wigs.add(_parseMap(
            data: (value),
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value == null) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: 'null',
              valueColor: nullColor,
              showComma: showComma,
            ),
          );
        }
      }
      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: '}', style: TextStyle(color: curlyBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }

  /// Creates a List View
  Widget _parseList({
    required List<dynamic> data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[]',
          valueColor: squareBracketColor,
          showComma: showComma,
        );
      } else {
        return Text('[]', style: TextStyle(color: squareBracketColor));
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}[',
            style: TextStyle(color: squareBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[',
          valueColor: squareBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int j = 0; j < data.length; j++) {
        final i = data[j];
        final showComma = j != data.length - 1;

        if (i is String) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: '"$i"',
            valueColor: stringColor,
            showComma: showComma,
          ));
        } else if (i is int) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: intColor,
            showComma: showComma,
          ));
        } else if (i is double) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: doubleColor,
            showComma: showComma,
          ));
        } else if (i is bool) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: boolColor,
            showComma: showComma,
          ));
        } else if (i is List) {
          wigs.add(_parseList(
            data: i,
            indent: newIndent,
            showComma: showComma,
          ));
        } else if (i is Map) {
          wigs.add(_parseMap(
              data: i as Map<String, dynamic>,
              indent: newIndent,
              showComma: showComma));
        } else if (i == null) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: 'null',
            valueColor: nullColor,
            showComma: showComma,
          ));
        }
      }

      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: ']', style: TextStyle(color: squareBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }
}

class _KeyValueWidget extends StatelessWidget {
  _KeyValueWidget({
    Key? key,
    required this.indent,
    this.jsonKey,
    required this.jsonValue,
    required this.valueColor,
    required this.showComma,
  }) : super(key: key);

  final int indent;
  final String? jsonKey;
  final String jsonValue;
  final Color valueColor;

  final Color keyColor = _globalKeyColor;

  final bool showComma;

  @override
  Widget build(BuildContext context) {
    return Text.rich(TextSpan(
      children: [
        TextSpan(text: _spaces(indent)),
        if (jsonKey != null)
          TextSpan(text: '"$jsonKey"', style: TextStyle(color: keyColor)),
        if (jsonKey != null)
          TextSpan(text: ' : ', style: TextStyle(color: _globalColonColor)),
        TextSpan(text: jsonValue, style: TextStyle(color: valueColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor)),
      ],
    ));
  }
}

/// Required to create indents
String _spaces(int multiple) {
  return ' ' * _globalTabLength * multiple;
}


lib/colored_json.dart
library colored_json;

import 'package:flutter/material.dart';
import 'dart:convert';

late int _globalTabLength;

late Color _globalCommaColor;

late Color _globalKeyColor;

late Color _globalColonColor;

class ColoredJson extends StatelessWidget {
  ColoredJson({
    Key? key,
    required this.data,
    this.backgroundColor = Colors.transparent,
    this.stringColor = Colors.green,
    this.nullColor = Colors.red,
    this.intColor = Colors.purple,
    this.boolColor = Colors.pink,
    this.doubleColor = Colors.deepPurple,
    this.curlyBracketColor = Colors.black,
    this.squareBracketColor = Colors.black,
    this.textStyle = const TextStyle(),
    this.indentLength = 4,
    this.keyColor = Colors.blue,
    this.commaColor = Colors.black,
    this.colonColor = Colors.deepPurple,
  }) : super(key: key) {
    _globalTabLength = indentLength;
    _globalCommaColor = commaColor;
    _globalColonColor = colonColor;
    _globalKeyColor = keyColor;
  }

  /// JSON String
  final String data;

  /// Background color of the Widget
  final Color backgroundColor;

  /// Color of value of type [String]
  final Color stringColor;

  /// Color of value of type null
  final Color nullColor;

  /// Color of value of type [int]
  final Color intColor;

  /// Color of value of type [bool]
  final Color boolColor;

  /// Color of value of type [double]
  final Color doubleColor;

  /// Color of Curly brackets
  final Color curlyBracketColor;

  /// Color of Square brackets
  final Color squareBracketColor;

  /// Textstyle of parsed json data
  final TextStyle textStyle;

  /// Indent length in spaces, default to 4
  final int indentLength;

  /// Color of JSON Color Key
  final Color keyColor;

  /// Color of comma
  final Color commaColor;

  /// Color of colon
  final Color colonColor;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: backgroundColor,
      textStyle: textStyle,
      child: _coloredView(
        jsonString: data,
        indent: 0,
      ),
    );
  }

  /// Takes JSON string and initial indent
  Widget _coloredView({
    required String jsonString,
    required int indent,
  }) {
    final decodedData = jsonDecode(jsonString);
    if (decodedData is Map) {
      return _parseMap(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else if (decodedData is List) {
      return _parseList(
        data: decodedData,
        indent: indent,
        showComma: false,
      );
    } else {
      throw 'Not Correct type only List or Map allowed';
    }
  }

  /// Creates a Map View
  Widget _parseMap({
    required Map data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.entries.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{}',
          valueColor: curlyBracketColor,
          showComma: showComma,
        );
      } else {
        return Text(
          '{}',
          style: TextStyle(color: curlyBracketColor),
        );
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}{',
            style: TextStyle(color: curlyBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '{',
          valueColor: curlyBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int i = 0; i < data.entries.length; i++) {
        final jKey = data.entries.toList()[i].key;
        final value = data.entries.toList()[i].value;

        final bool showComma = i != data.entries.length - 1;

        if (value is String) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: '"$value"',
                valueColor: stringColor,
                showComma: showComma),
          );
        } else if (value is int) {
          wigs.add(
            _KeyValueWidget(
                indent: newIndent,
                jsonKey: jKey,
                jsonValue: value.toString(),
                valueColor: intColor,
                showComma: showComma),
          );
        } else if (value is double) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: doubleColor,
              showComma: showComma,
            ),
          );
        } else if (value is bool) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: value.toString(),
              valueColor: boolColor,
              showComma: showComma,
            ),
          );
        } else if (value is List) {
          wigs.add(_parseList(
            data: value,
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value is Map) {
          wigs.add(_parseMap(
            data: (value),
            indent: newIndent,
            jkey: jKey,
            showComma: showComma,
          ));
        } else if (value == null) {
          wigs.add(
            _KeyValueWidget(
              indent: newIndent,
              jsonKey: jKey,
              jsonValue: 'null',
              valueColor: nullColor,
              showComma: showComma,
            ),
          );
        }
      }
      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: '}', style: TextStyle(color: curlyBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }

  /// Creates a List View
  Widget _parseList({
    required List<dynamic> data,
    required int indent,
    required bool showComma,
    String? jkey,
  }) {
    if (data.isEmpty) {
      if (jkey != null) {
        return _KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[]',
          valueColor: squareBracketColor,
          showComma: showComma,
        );
      } else {
        return Text('[]', style: TextStyle(color: squareBracketColor));
      }
    } else {
      final wigs = <Widget>[];

      if (jkey == null) {
        wigs.add(Text('${_spaces(indent)}[',
            style: TextStyle(color: squareBracketColor)));
      } else {
        wigs.add(_KeyValueWidget(
          indent: indent,
          jsonKey: jkey,
          jsonValue: '[',
          valueColor: squareBracketColor,
          showComma: false,
        ));
      }

      final int newIndent = indent + 1;

      for (int j = 0; j < data.length; j++) {
        final i = data[j];
        final showComma = j != data.length - 1;

        if (i is String) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: '"$i"',
            valueColor: stringColor,
            showComma: showComma,
          ));
        } else if (i is int) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: intColor,
            showComma: showComma,
          ));
        } else if (i is double) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: doubleColor,
            showComma: showComma,
          ));
        } else if (i is bool) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: i.toString(),
            valueColor: boolColor,
            showComma: showComma,
          ));
        } else if (i is List) {
          wigs.add(_parseList(
            data: i,
            indent: newIndent,
            showComma: showComma,
          ));
        } else if (i is Map) {
          wigs.add(_parseMap(
              data: i as Map<String, dynamic>,
              indent: newIndent,
              showComma: showComma));
        } else if (i == null) {
          wigs.add(_KeyValueWidget(
            indent: newIndent,
            jsonValue: 'null',
            valueColor: nullColor,
            showComma: showComma,
          ));
        }
      }

      wigs.add(Text.rich(TextSpan(children: [
        TextSpan(text: _spaces(indent)),
        TextSpan(text: ']', style: TextStyle(color: squareBracketColor)),
        if (showComma)
          TextSpan(text: ',', style: TextStyle(color: _globalCommaColor))
      ])));

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: wigs,
      );
    }
  }
}

class _KeyValueWidget extends StatelessWidget {
  _KeyValueWidget({
    Key? key,
    required this.indent,
    this.jsonKey,
    required this.jsonValue,
    required this.valueColor,
    required this.showComma,
  }) : super(key: key);

  final int indent;
  final String? jsonKey;
  final String jsonValue;
  final Color valueColor;

  final Color keyColor = _globalKeyColor;

  final bool showComma;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        SizedBox(
          child: Text(_spaces(indent)),
        ),
        Flexible(
          child: Text.rich(TextSpan(
            children: [
              if (jsonKey != null)
                TextSpan(text: '"$jsonKey"', style: TextStyle(color: keyColor)),
              if (jsonKey != null)
                TextSpan(
                    text: ' : ', style: TextStyle(color: _globalColonColor)),
              TextSpan(text: jsonValue, style: TextStyle(color: valueColor)),
              if (showComma)
                TextSpan(text: ',', style: TextStyle(color: _globalCommaColor)),
            ],
          )),
        ),
      ],
    );
  }
}

/// Required to create indents
String _spaces(int multiple) {
  return ' ' * _globalTabLength * multiple;
}


example/lib/main.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:colored_json/colored_json.dart';

void main() {
  runApp(const MyApp());
}

final dartMap = {
  "id": 24,
  "name": "Manthan Khandale",
  "score": 7.6,
  "socials": null,
  "hobbies": [
    "Music",
    "Filmmaking",
  ],
  "isFlutterCool": true,
};

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: const Text(
            'Colored Json Example',
          ),
        ),
        body: Column(
          children: [
            ColoredJson(
              data: jsonEncode(dartMap),
              indentLength: 5,
              keyColor: Colors.green,
              backgroundColor: Colors.black,
              boolColor: Colors.white,
              nullColor: Colors.redAccent,
              stringColor: Colors.cyan,
              curlyBracketColor: Colors.yellow,
              doubleColor: Colors.deepOrange,
              squareBracketColor: Colors.amber,
              commaColor: Colors.yellow,
              colonColor: Colors.purple,
              intColor: Colors.lime,
              textStyle: const TextStyle(
                fontSize: 20,
                letterSpacing: 2,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
